rts/motoko-rts/src/gc/mark_compact.rs

```diff
@@ -0,0 +1,312 @@
+//! Implements "threaded compaction" as described in The Garbage Collection Handbook section 3.3.
+
+use crate::bitmap::{alloc_bitmap, free_bitmap, get_bit, iter_bits, set_bit};
+use crate::closure_table::closure_table_loc;
+use crate::mark_stack::{self, alloc_mark_stack, free_mark_stack, pop_mark_stack};
+use crate::mem::memcpy_words;
+use crate::rts_trap_with;
+use crate::types::*;
+
+use super::{get_heap_base, get_static_roots, note_live_size, note_reclaimed, HP};
+
+#[no_mangle]
+unsafe extern "C" fn compacting_gc() {
+    let old_hp = HP;
+    let heap_base = get_heap_base();
+
+    mark_compact(heap_base, old_hp, get_static_roots(), closure_table_loc());
+
+    let reclaimed = old_hp - HP;
+    note_reclaimed(Bytes(reclaimed));
+
+    let new_live_size = HP - heap_base;
+    note_live_size(Bytes(new_live_size));
+}
+
+unsafe fn mark_compact(
+    heap_base: u32,
+    heap_end: u32,
+    static_roots: SkewedPtr,
+    closure_table_loc: *mut SkewedPtr,
+) {
+    let heap_size = Bytes(heap_end - heap_base);
+
+    alloc_bitmap(heap_size);
+    alloc_mark_stack();
+
+    mark_static_roots(static_roots, heap_base);
+
+    // TODO: We could skip the is_tagged_scalar, heap_base etc. checks
+    push_mark_stack(*closure_table_loc, heap_base);
+
+    mark_stack(heap_base);
+
+    thread_roots(static_roots, heap_base);
+    thread(closure_table_loc, heap_base);
+    update_fwd_refs(heap_base);
+    update_bwd_refs(heap_base);
+
+    free_bitmap();
+    free_mark_stack();
+}
+
+unsafe fn mark_static_roots(static_roots: SkewedPtr, heap_base: u32) {
+    let root_array = static_roots.as_array();
+
+    // Static objects are not in the dynamic heap so don't need marking.
+    for i in 0..root_array.len() {
+        let obj = root_array.get(i).unskew() as *mut Obj;
+        mark_fields(obj, heap_base);
```

Comment:
Why do we mark their fields? Because they may just be ref cells whose contents has been updated to point into the non-static heap?

Reply:
Yes. Else all stuff you put in your precious top-level `var state = …` will be lost :)

Reply:
So I think this part is a bit confusing. Currently static heap is just an array with `MutBox` elements. The `MutBox`es also live in the static heap so none of the array of the `MutBox`es need marking (or evacuation). However `MutBox`es point to dynamic heap so fields of `MutBox`es need to be marked. That's what we do in this code.

There's nothing else in the static heap currently, only the array and `MutBox`es.

You will probably ask "why not just store the MutBox fields in the array"? That's because the `MutBox`es can be referenced from other objects. We can't have interior pointers so if we remove the `MutBox`es and store fields directly in the array then we won't be able to refer to those fields directly, we will need to refer to the array, and then store the index of the reference for the field.

Reply:
> There's nothing else in the static heap currently, only the array and MutBoxes.

I don’t think that’s true; shouldn’t there be lots of constant values (texts etc.) in the static heap?

(The “static roots” are all MutBoxes assembled collected in one array, if that’s what you mean.)

Reply:
> I don’t think that’s true; shouldn’t there be lots of constant values (texts etc.) in the static heap?

Err, sorry, I should've said "as far as garbage collection concerned".

> And this doesn’t affect GC

What doesn't affect GC? The array is used by the GC as the root set, right?

Reply:
I mean, it doesn’t affect GC that there is other stuff in the static heap. I am just confirming your “as far as GC is concerned” :-)